### PR TITLE
Fix in checking channel_reserve

### PR DESCRIPTION
**Pull request by yuntai**

### CODE REVIEW

File: src/ln/channel.rs [hunk 1]
```diff
@@ -2609,8 +2609,8 @@ impl Channel {
 		if htlc_outbound_value_msat + amount_msat > self.their_max_htlc_value_in_flight_msat {
 			return Err(HandleError{err: "Cannot send value that would put us over our max HTLC value in flight", action: None});
 		}
-		// Check their_channel_reserve_satoshis:
-		if htlc_inbound_value_msat + htlc_outbound_value_msat + amount_msat + (self.channel_value_satoshis * 1000 - self.value_to_self_msat) > (self.channel_value_satoshis - self.their_channel_reserve_satoshis) * 1000 {
```
Comment (yuntai) [inline src/ln/channel.rs]: The previous eqn seems double counting htlc_inbound_value_msat and htlc_outbound_value_msat?

**Comment by TheBlueMatt** [comment]
Hmm, we definitely need to be subtracting pending HTLCs from the value_to_self_msat here (since any pending HTLCs can't count towards channel reserve and aren't included in value_to_self_msat), though it looks pretty awkward and I'd buy that its wrong. Can you add a test either way?

**Comment by yuntai** [comment]
I figured new conditions are the same except not including value for incoming HTLC to subtract from value_to_self_msat, so it's just tighter. Also, added outgoing value in the pending_cell.

**Comment by TheBlueMatt** [comment]
Hmm, ok, yes, I buy that the new version is correct, but really needs tests. One thing you could do it add a debug_assert in build_commitment_transaction that the finally-calculated value_to_self and value_to_remote match channel reserves (which is broader than revoke_and_ack), though I don't like relying on fuzz testing for that as the full_stack_target is still way too broad to reliably find things like that.

If you did that and also added a number of tests which tested sending one less than and one more than the limit in the ChannelManager network tests (incl various states like awaiting-revoke, etc) then we'd probably be OK.

Testing the holding cell updates is going to be much harder, I think, though, cause free_holding_cell_htlcs handles the error case instead of sending the new HTLC.

File: src/ln/channel.rs [hunk 2]
```diff
@@ -2614,8 +2614,19 @@ impl Channel {
...
+		for holding_htlc in self.holding_cell_htlc_updates.iter() {
+			match holding_htlc {
+				&HTLCUpdateAwaitingACK::AddHTLC { ref amount_msat, .. } => {
+					holding_cell_outbound_amount_msat += *amount_msat;
+				}
+				_ => {}
+			}
+		}
+
+		// Check self.their_channel_reserve_satoshis (i.e our channel reserve):
```
Comment (TheBlueMatt) [inline src/ln/channel.rs]: Not sure that that is any more clear...maybe s/our channel reserve/the amount we must keep as reserve for them to have something to claim if we misbehave/?

**Comment by yuntai** [comment]
still need another test involving holding cell.

**Comment by yuntai** [comment]
added a test scenario with holding cell.

File: src/ln/channelmanager.rs [hunk 3]
```diff
@@ -3827,6 +3835,285 @@ mod tests {
...
 	}
 
+	#[test]
+	fn channel_reserve_test() {
+		use util::rng;
+		use std::sync::atomic::Ordering;
+		use ln::msgs::HandleError;
+
+		macro_rules! get_channel_value_stat {
+      ($node: expr, $channel_id: expr) => {{
```
Comment (TheBlueMatt) [inline src/ln/channelmanager.rs]: nit: indentation is wrong around here.

File: src/ln/channelmanager.rs [hunk 4]
```diff
@@ -3827,6 +3835,285 @@ mod tests {
...
+		let mut stat01 = get_channel_value_stat!(nodes[0], chan_1.2);
+		let mut stat11 = get_channel_value_stat!(nodes[1], chan_1.2);
+
+		let mut stat12 = get_channel_value_stat!(nodes[1], chan_2.2);
+		let mut stat22 = get_channel_value_stat!(nodes[2], chan_2.2);
+
+		macro_rules! get_route_and_payment_hash {
+			($recv_value: expr) => {{
+				let route = nodes[0].router.get_route(&nodes.last().unwrap().node.get_our_node_id(), None, &Vec::new(), $recv_value, TEST_FINAL_CLTV).unwrap();
+				let payment_preimage = [*nodes[0].network_payment_count.borrow(); 32];
```
Comment (TheBlueMatt) [inline src/ln/channelmanager.rs]: Should be able to use get_payment_preimage_hash!!

**Comment by TheBlueMatt** [comment]
You can also take the change at https://github.com/TheBlueMatt/rust-lightning/commit/1f31e03fac5d451d2cc47a06d681bcd73e4692a2 if you want, its a pretty aggressive check but once it gets adapted to udpate_fee/etc it may make sense to have something more aggressive like that.

**Comment by TheBlueMatt** [comment]
Will take as #206.

**Pull request closed without merging**